### PR TITLE
Feature/bicgstab

### PR DESCRIPTION
**Pull request by sunilkpai**
This is an update on #3796 for an implementation of BICGSTAB for the new repo and is no longer a WIP. I used the same tests as GMRES and they seem to work fine.

### CODE REVIEW

File: jax/_src/scipy/sparse/linalg.py
Comment (ianwilliamson) [inline jax/_src/scipy/sparse/linalg.py:179]: You could tree map `ones_like` I think?

Comment (sunilkpai) [reply jax/_src/scipy/sparse/linalg.py:179]: Here I am just trying to get the number 1 to match the dtype (complex or float, 32 or 64 bit). How would that work?

Comment (ianwilliamson) [reply jax/_src/scipy/sparse/linalg.py:179]: Sorry... thought I was being helpful with my drive by comment... so you actually want a scalar `1` . Perhaps https://jax.readthedocs.io/en/latest/_autosummary/jax.numpy.result_type.html#jax.numpy.result_type would be helpful, but I'll defer to someone else who knows for sure!

File: tests/lax_scipy_sparse_test.py
Comment (sunilkpai) [reply tests/lax_scipy_sparse_test.py:292]: @shoyer By just adding this line, both GMRES and BICGSTAB seem to fail this test, moreso for the `complex64` and `float32` dtypes. But since `scipy` matches, I expect the same thing would happen if we had used `scipy` for the iterative solves.

Comment (shoyer) [reply tests/lax_scipy_sparse_test.py:295]: Try relaxing the error tolerance a little bit?

Comment (sunilkpai) [reply tests/lax_scipy_sparse_test.py:295]: Hmm, ok it's already 0.2 right now, and I saw this case for GMRES on my local tests:

```
Mismatch: 100%
Max absolute difference: 0.6610913
Max relative difference: 161.79356
 x: array(0.657005, dtype=float32)
 y: array(-0.004086, dtype=float32)
```

Comment (shoyer) [reply tests/lax_scipy_sparse_test.py:295]: Hmm! My guess is that there is a genuine bug in the derivative rule for GMRES. I would suggest commenting out this test for GMRES for now (opening an issue so we don't forget about it), and we can try to fix that in a follow-on PR.

**Comment by sunilkpai** [comment]
@shoyer It's ready now. I needed to relax the gradient check a bit more to 0.3 for BICGSTAB, and I think we can address GMRES gradient issues in a future PR as you mentioned so I commented it out for now.

**Comment by shoyer** [comment]
Unfortunately some of these gradient checks (both GMRES and BICGSTAB) are failing on Google's internal CI, which also runs things on GPU/TPU and Google's slightly different linear algebra libraries for CPU.

This is rather annoying for non-Googlers to debug, but I suspect it's entirely spurious because it mostly turns up for low precision. If you want to give this one more try, perhaps set `order=1` in `check_grads()` (under the theory that higher order derivatives have larger numerical error). Otherwise, I guess we can comment out those checks, too, for now.

**Comment by sunilkpai** [comment]
@shoyer what do you think is the best course of action here?

**Comment by shoyer** [comment]
One of the GitHub CI tests seems to be timing out. It used to run in about 30 minutes and is now taking over an hour. Can you check to see if the tests this PR introduces are much slower than expected? 

The timing out test sets `JAX_NUM_GENERATED_CASES=25`:
https://jax.readthedocs.io/en/latest/developer.html#running-the-tests

**Comment by shoyer** [comment]
It might be worth squashing your changes and rebasing on master to see if that helps the slow tests.

**Comment by sunilkpai** [comment]
Hi @shoyer hmm no dice on the squashing. I couldn't immediately find which tests took so long (it didn't seem to matter if I increase the number of cases), but the longest tests seemed to be whereever grads are checked for CG. Is there a way I can see which tests took so long during CI?

**Comment by shoyer** [comment]
I thought you commented out the gradient tests? If you can't reproduce this locally I'm honestly baffled here. Let's see if other JAX maintainers have ideas...

**Comment by shoyer** [comment]
If you could sync in the latest changes from the JAX repo, the unit tests will run in verbose mode. That should hopefully make it more obvious where the test suite is getting stuck.

**Comment by sunilkpai** [comment]
@shoyer re: grad tests, I only kept the grad tests that didn't pose an issue in master! And will do!

**Comment by sunilkpai** [comment]
@shoyer I'm wondering if the issue was because of the x64 enable I accidentally kept?

File: tests/lax_scipy_sparse_test.py
Comment (shoyer) [inline tests/lax_scipy_sparse_test.py:33]: Good catch! This was probably it. Can you remove this commented out line entirely?

Comment (sunilkpai) [reply tests/lax_scipy_sparse_test.py]: Done!

**Comment by sunilkpai** [comment]
@shoyer I think this should be good to go, let me know if there's anything else you'd like me to change!

**Review by shoyer — APPROVED** [review]

**Comment by shoyer** [comment]
Thanks for your perseverance here, @sunilkpai !

We are really excited to have this feature.

**Comment by sunilkpai** [comment]
Awesome, thanks @shoyer, happy to contribute!

**Comment by tetterl** [comment]
Great work! For visibility https://jax.readthedocs.io/en/latest/jax.scipy.html#module-jax.scipy.sparse.linalg should be updated as well.

**Comment by jakevdp** [comment]
Thanks! Done in https://github.com/google/jax/pull/6647

**Pull request merged**
